src/error/error.h

```diff
@@ -0,0 +1,199 @@
+#pragma once
+
+#include <string>
+#include <stdexcept>
+
+// Error utilities.
+
+/// Base error interface.
+/// All error classes must derive from this one.
+class Error {
+ public:
+  /// Conversion to a string for display.
+  virtual std::string toString() const = 0;
+  virtual ~Error() {}
+};
+
+/// Basic error, with an error message.
+class GenericError : public Error {
+ public:
+  /// Build a new generic error, with an error message.
+  explicit GenericError(const std::string& message) : Error(), message_(message) {}
+
+  /// Return the error message it was constructed with.
+  virtual std::string toString() const override { return message(); }
```

Comment:
We usually don't put both virtual and override: override already indicates this method is virtual.